src/csharp/Grpc.IntegrationTesting/XdsInteropClient.cs

```diff
@@ -0,0 +1,300 @@
+        internal async Task RunChannelsAsync(CancellationToken cancellationToken)
+        {
+            var channelTasks = new List<Task>();
+            for (int channelId = 0; channelId < options.NumChannels; channelId++)
+            {
+                var channelTask = RunSingleChannelAsync(channelId, cancellationToken);
+                channelTasks.Add(channelTask);
+            }
+
+            for (int channelId = 0; channelId < options.NumChannels; channelId++)
+            {
+                await channelTasks[channelId];
+            }
+        }
+
+        private async Task RunSingleChannelAsync(int channelId, CancellationToken cancellationToken)
+        {
+            Console.WriteLine($"Starting channel {channelId}");
+            var channel = new Channel(options.Server, ChannelCredentials.Insecure);
+            var client = new TestService.TestServiceClient(channel);
+
+            var inflightTasks = new List<Task>();
+            long rpcsStarted = 0;
+            var stopwatch = Stopwatch.StartNew();
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                inflightTasks.Add(RunSingleRpcAsync(client, cancellationToken));
+                rpcsStarted++;
+
+                // only cleanup calls that have already completed, calls that are still inflight will be cleaned up later.
+                await CleanupCompletedTasksAsync(inflightTasks);
+
+                Console.WriteLine($"Currently {inflightTasks.Count} in-flight RPCs");
+
+                // if needed, wait a bit before we start the next RPC.
+                int nextDueInMillis = (int) Math.Max(0, (1000 * rpcsStarted / options.Qps) - stopwatch.ElapsedMilliseconds);
+                if (nextDueInMillis > 0)
+                {
+                    await Task.Delay(nextDueInMillis);
+                }
+            }
+            stopwatch.Stop();
+
+            Console.WriteLine($"Shutting down channel {channelId}");
+            await channel.ShutdownAsync();
+            Console.WriteLine($"Channel shutdown {channelId}");
+        }
+
+        private async Task RunSingleRpcAsync(TestService.TestServiceClient client, CancellationToken cancellationToken)
+        {
+            long rpcId = statsWatcher.RpcIdGenerator.Increment();
+            try
+            {
+                Console.WriteLine($"Starting RPC {rpcId}.");
+                var response = await client.UnaryCallAsync(new SimpleRequest(),
+                    new CallOptions(cancellationToken: cancellationToken, deadline: DateTime.UtcNow.AddSeconds(options.RpcTimeoutSec)));
+                
+                statsWatcher.OnRpcComplete(rpcId, response.Hostname);
+                if (options.PrintResponse)
+                {
+                    Console.WriteLine($"Got response {response}");
+                }
+                Console.WriteLine($"RPC {rpcId} succeeded ");
+            }
+            catch (RpcException ex)
+            {
```

Comment:
in the [spec](https://github.com/grpc/grpc/blob/master/doc/xds-test-descriptions.md) there's an option to "fail on failed RPCs". But AFAICS we're not consuming that flag here.

Should the code in the `catch` block here be checking that flag, and raising the error up, if set true?

If this isn't important for now, but planned for later, then maybe just add a TODO?

Reply:
I actually added the `--fail_on_failed_rpcs` flag to the spec after this PR was started. If it's easy enough to add support for it here that would be good, but a TODO to add it later would be fine too (the `run_xds_tests.py` test runner is backwards compatible for clients that don't implement this flag).